qa/rpc-tests/finalsaplingroot.py

```diff
@@ -0,0 +1,131 @@
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # Verify the final Sapling root has changed
+        blk = self.nodes[0].getblock("201")
+        root = blk["finalsaplingroot"]
+        assert(root is not SAPLING_TREE_EMPTY_ROOT)        
+        assert(root is not NULL_FIELD)  
+
+        # Verify there is a Sapling output description (its commitment was added to tree)
+        result = self.nodes[0].getrawtransaction(mytxid, 1)
+        assert_equal(len(result["vShieldedOutput"]), 1)
+
+        # Mine an empty block and verify the final Sapling root does not change
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+        assert_equal(root, self.nodes[0].getblock("202")["finalsaplingroot"])
+
+        # Mine a block with a transparent tx and verify the final Sapling root does not change
+        taddr1 = self.nodes[1].getnewaddress()
+        self.nodes[0].sendtoaddress(taddr1, Decimal("1.23"))
+
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        assert_equal(len(self.nodes[0].getblock("203")["tx"]), 2)
+        assert_equal(self.nodes[1].z_getbalance(taddr1), Decimal("1.23"))
+        assert_equal(root, self.nodes[0].getblock("203")["finalsaplingroot"])
+
+        # Mine a block with a Sprout shielded tx and verify the final Sapling root does not change
+        zaddr1 = self.nodes[1].z_getnewaddress()
+        recipients = []
+        recipients.append({"address": zaddr1, "amount": Decimal('10')})
+        myopid = self.nodes[0].z_sendmany(taddr0, recipients, 1, 0)
+        wait_and_assert_operationid_status(self.nodes[0], myopid)
+
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        assert_equal(len(self.nodes[0].getblock("204")["tx"]), 2)
+        assert_equal(self.nodes[1].z_getbalance(zaddr1), Decimal("10"))
+        assert_equal(root, self.nodes[0].getblock("204")["finalsaplingroot"])
+
+        # Mine a block with a Sapling shielded tx and verify the final Sapling root changes
+        saplingAddr1 = self.nodes[1].z_getnewaddress("sapling")
+        recipients = []
+        recipients.append({"address": saplingAddr1, "amount": Decimal('12.34')})
+        myopid = self.nodes[0].z_sendmany(saplingAddr0, recipients, 1, 0)
+        mytxid = wait_and_assert_operationid_status(self.nodes[0], myopid)
+
+        self.sync_all()
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        assert_equal(len(self.nodes[0].getblock("205")["tx"]), 2)
+        assert_equal(self.nodes[1].z_getbalance(saplingAddr1), Decimal("12.34"))
+        assert(root is not self.nodes[0].getblock("205")["finalsaplingroot"])
+
+        # Verify there is a Sapling output description (its commitment was added to tree)
+        result = self.nodes[0].getrawtransaction(mytxid, 1)
+        assert_equal(len(result["vShieldedOutput"]), 2)  # there is Sapling shielded change
+
+if __name__ == '__main__':
```

Comment:
Nice job.
Might be nice to add a test where the sender is a shielded spend but not recipient
and see the root doesn't change. Perhaps I'll add a commit for that.

Reply:
ACK